File: scraper/scraper.py
import requests
from bs4 import BeautifulSoup
import json
import re
import time
from urllib.parse import urljoin

BASE_URL = "https://toonitalia.xyz"
HEADERS = {
    "User-Agent": "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/125.0.0.0 Safari/537.36",
    "Accept-Language": "it-IT,it;q=0.9,en-US;q=0.8,en;q=0.7",
}

session = requests.Session()
session.headers.update(HEADERS)


def get_soup(url):
    resp = session.get(url, timeout=30)
    resp.raise_for_status()
    return BeautifulSoup(resp.text, "lxml")


def scrape_categories():
    categories = {
        "anime-ita": "Anime Ita",
        "contatti": "Anime Sub-Ita",
        "film-animazione": "Film Animazione",
        "serie-tv": "Serie Tv",
    }
    return categories


def scrape_content_list(category_slug):
    url = f"{BASE_URL}/{category_slug}/"
    soup = get_soup(url)
    items = []
    seen = set()

    # Category pages list every title in .catlist-box > ul > li > a[href].
    for link in soup.select(".catlist-box ul li a[href]"):
        href = link.get("href", "")
        title = link.get_text(strip=True)
        if title and BASE_URL in href and href not in seen:
            seen.add(href)
            items.append({
                "title": title,
                "url": href,
                "image": "",
                "category": category_slug,
            })
    return items


def scrape_homepage():
    soup = get_soup(BASE_URL)
    sections = {}
    current_section = None

    # The homepage is laid out as a series of .col blocks, each with an <h2>
    # section title and a list of .item cards (.card-link > img + .title span).
    for col in soup.select(".col"):
        h2 = col.find("h2")
        if not h2:
            continue
        text = h2.get_text(strip=True)
        # strip a leading emoji / decorative char if present
        text = re.sub(r"^[\u0000-\u001F\u200B-\u200D\uFEFF\u20E3\u2600-\u27BF\U0001F000-\U0001FFFF]\s*", "", text).strip()
        if not text:
            continue
        current_section = text
        sections[current_section] = []

        for item in col.select(".item"):
            link = item.find("a", href=True)
            if not link:
                continue
            href = link["href"]
            img = item.find("img")
            image_url = img["src"] if img and img.get("src") else ""
            title_el = item.find(class_="title")
            title = title_el.get_text(strip=True) if title_el else link.get_text(strip=True)
            if title and BASE_URL in href and href != BASE_URL:
                sections[current_section].append({
                    "title": title,
                    "url": href,
                    "image": image_url,
                })

    # Fallback: if the layout changed, build sections from the category pages.
    if not sections:
        for slug, name in scrape_categories().items():
            try:
                items = scrape_content_list(slug)[:30]
                if items:
                    sections[name] = items
            except Exception:
                pass

    return sections


HOSTS = [
    "chuckle-tube.com", "vidhideplus.com", "uqload.is", "uqload.com",
    "uqload.bz", "uqload.to", "voe.sx", "ryderjet.com", "luluvdo.com",
    "streamtape", "strcloud.link", "scloud.lol", "supervideo", "dood",
    "mixdrop", "filelions", "streamwish", "vidmoly", "embedsb",
]


def _is_streaming_link(href):
    return any(h in href for h in HOSTS)


def scrape_detail(url):
    soup = get_soup(url)
    result = {"url": url, "episodes": [], "seasons": []}

    title_tag = soup.find("h1")
    result["title"] = title_tag.get_text(strip=True) if title_tag else ""

    content = soup.find("article") or soup.find("div", class_="entry-content") or soup
    text = content.get_text("\n", strip=True)

    for field, key in [
        ("Titolo originale", "original_title"),
        ("Paese di origine", "country"),
        ("Data di pubblicazione", "year"),
        ("Stato Opera", "status"),
        ("N. Episodi", "total_episodes"),
        ("Episodi disponibili", "available_episodes"),
    ]:
        m = re.search(rf"{field}:\s*(.+)", text)
        if m:
            result[key] = m.group(1).strip()

    img_tag = content.find("img")
    if img_tag and img_tag.get("src"):
        result["image"] = img_tag["src"]

    trama_match = re.search(r"Trama:\s*(.+?)(?=\n\n|Company|Episodi|Stagione)", text, re.DOTALL)
    if trama_match:
        result["synopsis"] = trama_match.group(1).strip()

    current_season = 0
    episodes = []
    # Episode blocks live inside <p> elements (often nested inside a div), so we
    # walk every <p> descendant, not just direct children.
    for el in content.find_all("p"):
        # Track season headers that appear before this paragraph.
        prev = el.find_previous(["h2", "h3"])
        if prev is not None:
            sm = re.search(r"(\d+)\s*[°º]\s*Stagione", prev.get_text())
            if sm:
                current_season = int(sm.group(1))

        for part_html in re.split(r"<br\s*/?>", str(el), flags=re.IGNORECASE):
            seg = BeautifulSoup(f"<p>{part_html}</p>", "lxml").find("p")
            if not seg:
                continue
            links = seg.find_all("a", href=True)
            if not links:
                continue
            if not any(_is_streaming_link(a["href"]) for a in links):
                continue

            player_links = []
            for a in links:
                href = a["href"]
                if not href:
                    continue
                label = a.get_text(strip=True) or "Player"
                player_links.append({"label": label, "url": href})
            if not player_links:
                continue

            seg_text = seg.get_text()
            clean_text = seg_text
            for a in links:
                clean_text = clean_text.replace(a.get_text(), "").strip()
            clean_text = re.sub(r"\s*[–\-–]\s*$", "", clean_text).strip()
            # Drop a trailing separator that may remain after the last link.
            clean_text = re.sub(r"[–\-–]\s*$", "", clean_text).strip()

            m1 = re.match(r"^(\d+)\s*[–-]\s*(\d{4})\s*[–-]\s*(.+?)\s*$", clean_text)
            m2 = re.match(r"^(\d+)\s*[–-]\s*(.+?)\s*$", clean_text)
            ep_num = int(m1.group(1)) if m1 else (int(m2.group(1)) if m2 else (len(episodes) + 1))
            ep_title = (m1.group(3).strip() if m1 else (m2.group(2).strip() if m2 else f"Episodio {ep_num}"))

            episodes.append({
                "title": f"Ep. {ep_num} - {ep_title}",
                "url": player_links[0]["url"],
                "season": current_season,
                "number": ep_num,
                "players": player_links,
            })

    result["episodes"] = episodes

    seasons = []
    for s in content.select("[id^=S], a[name^=S]"):
        sid = s.get("id") or s.get("name")
        if sid:
            seasons.append(sid)
    result["seasons"] = list(dict.fromkeys(seasons))

    return result


def scrape_all_content():
    all_content = []
    categories = scrape_categories()

    for slug, name in categories.items():
        print(f"Scraping {name}...")
        items = scrape_content_list(slug)
        for item in items:
            print(f"  -> {item['title']}")
            detail = scrape_detail(item["url"])
            detail["category"] = name
            detail["category_slug"] = slug
            detail["thumbnail"] = item.get("image", "")
            all_content.append(detail)
            time.sleep(0.5)

    return all_content


def save_json(data, filename="toonitalia_data.json"):
    with open(filename, "w", encoding="utf-8") as f:
        json.dump(data, f, ensure_ascii=False, indent=2)
    print(f"Saved {len(data)} items to {filename}")


if __name__ == "__main__":
    print("=== ToonItalia Scraper ===")
    print("1. Scrape homepage")
    print("2. Scrape all content")
    print("3. Scrape single URL")
    choice = input("Choice (1/2/3): ").strip()

    if choice == "1":
        data = scrape_homepage()
        save_json(data, "toonitalia_homepage.json")
        print(json.dumps(data, indent=2, ensure_ascii=False)[:3000])

    elif choice == "2":
        data = scrape_all_content()
        save_json(data)

    elif choice == "3":
        url = input("URL: ").strip()
        data = scrape_detail(url)
        save_json([data])
        print(json.dumps(data, indent=2, ensure_ascii=False)[:3000])

File: android/app/src/main/java/com/toonitalia/app/Scraper.kt
package com.toonitalia.app

import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.async
import kotlinx.coroutines.awaitAll
import kotlinx.coroutines.coroutineScope
import kotlin.text.RegexOption
import org.jsoup.Jsoup
import java.net.URLEncoder

object Scraper {
    private const val BASE_URL = "https://toonitalia.xyz"

    private fun getDoc(url: String): org.jsoup.nodes.Document {
        val html = NetworkModule.fetch(url)
        return Jsoup.parse(html)
    }

    private var homepageImageCache = mutableMapOf<String, String>()

    private fun buildHomepageCache() {
        if (homepageImageCache.isNotEmpty()) return
        try {
            val doc = getDoc(BASE_URL)
            for (item in doc.select(".item")) {
                val link = item.selectFirst("a[href]") ?: continue
                val href = link.attr("href")
                val img = item.selectFirst("img")
                val imageUrl = img?.attr("src") ?: ""
                if (href.isNotBlank() && imageUrl.isNotBlank()) {
                    homepageImageCache[href.trimEnd('/')] = imageUrl
                }
            }
        } catch (_: Exception) {}
    }

    private fun getImageForUrl(url: String): String {
        return homepageImageCache[url.trimEnd('/')] ?: ""
    }

    private fun cleanEmojiTitle(raw: String?): String {
        if (raw == null) return ""
        return raw
            .replace(Regex("[\\u0000-\\u001F\\u200B-\\u200D\\uFEFF]"), "")
            .replace(Regex("[\\u20E3\\u2600-\\u27BF\\u{1F000}-\\u{1FAFF}\\u{2600}-\\u{27BF}]"), "")
            .trim()
    }

    suspend fun scrapeHomepage(): List<CategorySection> {
        return try {
            val doc = getDoc(BASE_URL)
            val sections = mutableListOf<CategorySection>()

            val cols = doc.select(".col")
            for (col in cols) {
                val h2 = col.selectFirst("h2")
                val title = cleanEmojiTitle(h2?.text()) ?: continue
                if (title.isBlank()) continue

                val items = mutableListOf<ContentItem>()
                for (item in col.select(".item")) {
                    val link = item.selectFirst("a[href]")
                    val href = link?.attr("href") ?: continue
                    val img = item.selectFirst("img")
                    val imageUrl = img?.attr("src") ?: ""
                    val itemTitle = item.selectFirst(".title")?.text()?.trim()
                        ?: link.text().trim()

                    if (itemTitle.isNotBlank() && href.startsWith(BASE_URL) && href != BASE_URL) {
                        items.add(ContentItem(
                            title = itemTitle,
                            url = href,
                            image = imageUrl
                        ))
                    }
                }
                if (items.isNotEmpty()) {
                    sections.add(CategorySection(title = title, items = items))
                }
            }

            if (sections.isNotEmpty()) {
                // Cache images for search/category lookups
                for (section in sections) {
                    for (item in section.items) {
                        if (item.image.isNotBlank()) homepageImageCache[item.url.trimEnd('/')] = item.image
                    }
                }
                sections
            } else {
                fallbackHomepage()
            }
        } catch (e: Exception) {
            e.printStackTrace()
            fallbackHomepage()
        }
    }

    private suspend fun fallbackHomepage(): List<CategorySection> {
        val cats = scrapeCategories()
        val sections = mutableListOf<CategorySection>()
        for ((slug, label) in cats) {
            try {
                val items = scrapeContentList(slug).take(30)
                if (items.isNotEmpty()) {
                    sections.add(CategorySection(title = label, items = items))
                }
            } catch (_: Exception) {
            }
        }
        return sections
    }

    suspend fun scrapeContentList(categorySlug: String): List<ContentItem> {
        val doc = getDoc("$BASE_URL/$categorySlug/")
        val items = mutableListOf<ContentItem>()
        val seen = mutableSetOf<String>()

        for (a in doc.select(".catlist-box ul li a[href]")) {
            val href = a.attr("href")
            val title = a.text().trim()

            if (title.isNotBlank() && href.startsWith(BASE_URL) && !seen.contains(href)) {
                seen.add(href)
                items.add(ContentItem(
                    title = title,
                    url = href,
                    categorySlug = categorySlug
                ))
            }
        }

        buildHomepageCache()

        for (item in items) {
            if (item.image.isEmpty()) {
                val cached = getImageForUrl(item.url)
                if (cached.isNotEmpty()) {
                    items[items.indexOf(item)] = item.copy(image = cached)
                }
            }
        }

        val needsFetch = items.filter { it.image.isEmpty() }.take(30)
        if (needsFetch.isNotEmpty()) {
            try {
                val fetched = runCatching {
                    coroutineScope {
                        needsFetch.chunked(6).map { chunk ->
                            chunk.map { item ->
                                async(Dispatchers.IO) {
                                    try {
                                        val detailDoc = getDoc(item.url)
                                        val img = detailDoc.selectFirst("article .entry-content img")
                                            ?: detailDoc.selectFirst("article img")
                                            ?: detailDoc.selectFirst(".entry-content img")
                                        val imgUrl = img?.attr("src") ?: ""
                                        if (imgUrl.isNotBlank()) item to imgUrl else null
                                    } catch (_: Exception) { null }
                                }
                            }.awaitAll().filterNotNull()
                        }.flatten()
                    }
                }.getOrDefault(emptyList())

                for ((item, imgUrl) in fetched) {
                    val idx = items.indexOfFirst { it.url == item.url }
                    if (idx >= 0) {
                        items[idx] = items[idx].copy(image = imgUrl)
                    }
                }
            } catch (_: Exception) {}
        }

        return items
    }

    private fun isStreamingHref(href: String?): Boolean {
        if (href.isNullOrBlank() || !href.contains("://")) return false
        if (href.contains("toonitalia.xyz")) return false
        return true
    }

    private fun hostLabel(href: String): String {
        return try {
            val host = Regex("https?://([^/]+)/").find(href)?.groupValues?.get(1) ?: href
            host.replace("www.", "")
        } catch (_: Exception) { "Player" }
    }

    private fun isAdText(text: String): Boolean {
        return text.contains("http")
    }

    private fun parseEpisodeSegment(segment: String, seasonLabel: String, fallbackNumber: Int): Episode? {
        // segment looks like: "929 – Titolo episodio – PLAYER1 – PLAYER2" or "VOE – VIDHIDE" (film)
        val segDoc = Jsoup.parse("<p>$segment</p>")
        val segEl = segDoc.selectFirst("p") ?: return null
        val links = segEl.select("a[href]").filter { a -> isStreamingHref(a.attr("href")) }
        if (links.isEmpty()) return null

        val playerLinks = links.map { a ->
            val href = a.attr("href")
            val label = a.text().trim().ifBlank { hostLabel(href) }
            PlayerLink(label = label, url = href)
        }

        // Remove link texts from the segment text to isolate the episode title.
        var text = segEl.text()
        for (a in links) text = text.replace(a.text(), " ")
        text = text.replace(Regex("\\s+"), " ").trim()
        text = text.replace(Regex("^[^\\d]*?(\\d{2,4})\\s*[–-]\\s*"), "$1 – ").trim()
        text = text.replace(Regex("[–-]\\s*$"), "").trim()

        val numMatch = Regex("^(\\d{2,4})").find(text)
        val epNum = numMatch?.groupValues?.get(1)?.toIntOrNull() ?: fallbackNumber
        val epTitle = if (numMatch != null) {
            text.removePrefix(numMatch.groupValues[1]).replace(Regex("^\\s*[–-]\\s*"), "").trim()
        } else {
            text.replace(Regex("link\\s*streaming\\s*:", RegexOption.IGNORE_CASE), "")
                .replace(Regex("^[\\s–-]+"), "")
                .trim()
        }.ifBlank { seasonLabel.ifBlank { "Streaming" } }

        return Episode(
            title = if (numMatch != null) "Ep. $epNum - $epTitle" else epTitle,
            url = playerLinks.first().url,
            season = 0,
            number = epNum,
            players = playerLinks
        )
    }

    private fun splitEpisodeSegments(raw: String): List<String> {
        // Episodes are separated by " NNN - Title - PLAYER ..." on the same line.
        val parts = raw.split(Regex("(?<=\\s)(?=\\d{2,4}\\s*[–-]\\s)"))
        return parts.map { it.trim() }.filter { it.isNotBlank() }
    }

    fun scrapeDetail(url: String): ContentItem {
        val doc = getDoc(url)
        val episodes = mutableListOf<Episode>()

        val titleEl = doc.selectFirst("h1")
        val title = titleEl?.text()?.trim() ?: ""

        val content = doc.selectFirst("article .entry-content") ?: doc.selectFirst("article") ?: doc

        val img = content.selectFirst("img")
        val imageUrl = img?.attr("src") ?: ""

        // --- Metadata (single <p> with "Field : value Field2 : value2") ---
        var originalTitle = ""
        var country = ""
        var year = ""
        var status = ""
        var totalEpisodes = ""
        var availableEpisodes = ""
        for (p in content.select("p")) {
            val t = p.text()
            if ("Titolo originale" in t || "Paese di origine" in t || "Data di pubblicazione" in t
                || "Stato Opera" in t || "N. Episodi" in t || "Aggiornamento episodi" in t
            ) {
                for ((field, key) in mapOf(
                    "Titolo originale" to "originalTitle",
                    "Titoli alternativi" to "originalTitle",
                    "Paese di origine" to "country",
                    "Data di pubblicazione" to "year",
                    "Stato Opera" to "status",
                    "N. Episodi" to "totalEpisodes",
                    "Aggiornamento episodi" to "availableEpisodes"
                )) {
                    val m = Regex("$field\\s*:\\s*([^\\n]+?)(?=\\s{2,}[A-Z][^:]*:|$)").find(t)
                    if (m != null) {
                        val v = m.groupValues[1].trim().trimEnd('.').trim()
                        when (key) {
                            "originalTitle" -> if (originalTitle.isBlank()) originalTitle = v
                            "country" -> if (country.isBlank()) country = v
                            "year" -> if (year.isBlank()) year = v
                            "status" -> if (status.isBlank()) status = v
                            "totalEpisodes" -> if (totalEpisodes.isBlank()) totalEpisodes = v
                            "availableEpisodes" -> if (availableEpisodes.isBlank()) availableEpisodes = v
                        }
                    }
                }
            }
        }

        // --- Synopsis (between "Trama:" and the next relevant heading) ---
        var synopsis = ""
        var grab = false
        for (el in content.select("> *")) {
            val name = el.tagName()
            if (name == "h3" || name == "h2") {
                val txt = el.text()
                if ("Trama" in txt) { grab = true; continue }
                if (grab) break
            }
            if (grab && name == "p") {
                val t = el.text().trim()
                if (t.isNotBlank() && "PLAYER" !in t && !isAdText(t)) {
                    synopsis = t
                    break
                }
            }
        }

        // --- Episodes: walk direct children, split multi-episode <p> blocks ---
        var currentSeason = ""
        var fallbackCounter = 1
        for (el in content.select("> *")) {
            val name = el.tagName()
            if (name == "h3" || name == "h2") {
                val txt = el.text().trim()
                if (Regex("\\d+\\s*[°º]\\s*Stagione", RegexOption.IGNORE_CASE).containsMatchIn(txt)) {
                    currentSeason = txt
                } else if ("Sigle" in txt || "Scegli Stagione" in txt || "Special" in txt) {
                    currentSeason = txt
                }
                continue
            }
            if (name != "p") continue

            val links = el.select("a[href]").filter { a -> isStreamingHref(a.attr("href")) }
            if (links.isEmpty()) continue

            val raw = el.text()
            val segments = splitEpisodeSegments(raw)
            if (segments.isEmpty()) {
                // Film / single block without episode numbers
                val ep = parseEpisodeSegment(raw, currentSeason.ifBlank { title }, fallbackCounter++)
                if (ep != null) episodes.add(ep)
            } else {
                for (seg in segments) {
                    val ep = parseEpisodeSegment(seg, currentSeason.ifBlank { title }, fallbackCounter++)
                    if (ep != null) episodes.add(ep)
                }
            }
        }

        return ContentItem(
            title = title,
            url = url,
            image = imageUrl,
            synopsis = synopsis,
            originalTitle = originalTitle,
            country = country,
            year = year,
            status = status,
            totalEpisodes = totalEpisodes,
            availableEpisodes = availableEpisodes,
            episodes = episodes,
            seasons = emptyList()
        )
    }

    fun scrapeCategories(): Map<String, String> {
        return mapOf(
            "anime-ita" to "Anime Ita",
            "contatti" to "Anime Sub-Ita",
            "film-animazione" to "Film Animazione",
            "serie-tv" to "Serie Tv"
        )
    }

    suspend fun search(query: String): List<ContentItem> {
        val results = mutableListOf<ContentItem>()
        try {
            val encoded = URLEncoder.encode(query, "UTF-8")
            val doc = getDoc("$BASE_URL/?s=$encoded")
            buildHomepageCache()
            val rawItems = mutableListOf<ContentItem>()
            for (article in doc.select("article")) {
                val h2 = article.selectFirst("h2")
                val link = h2?.selectFirst("a[href]") ?: continue
                val href = link.attr("href")
                val title = link.text().trim()
                if (title.isNotBlank() && href.startsWith(BASE_URL) &&
                    !href.contains("/?s=") && !href.contains("/category/") &&
                    !href.contains("/author/") && !href.contains("/tag/")) {
                    // Search result cards have no inline image: try the homepage cache first.
                    var imageUrl = getImageForUrl(href)
                    if (imageUrl.isBlank()) {
                        val img = article.selectFirst("img")
                        imageUrl = img?.attr("src") ?: ""
                        if (imageUrl.isBlank()) {
                            val schemaScript = article.selectFirst("script[type=application/ld+json]")
                            if (schemaScript != null) {
                                val text = schemaScript.html()
                                val thumbMatch = Regex("\"thumbnailUrl\"\\s*:\\s*\"([^\"]+)\"").find(text)
                                imageUrl = thumbMatch?.groupValues?.get(1) ?: ""
                            }
                        }
                        if (imageUrl.isBlank()) {
                            val metaImg = article.selectFirst("meta[property=og:image]")
                            imageUrl = metaImg?.attr("content") ?: ""
                        }
                    }
                    rawItems.add(ContentItem(title = title, url = href, image = imageUrl))
                }
            }
            results.addAll(rawItems)

            // Fetch detail pages (in parallel) only for items still missing an image.
            val needsFetch = rawItems.filter { it.image.isBlank() }.take(20)
            if (needsFetch.isNotEmpty()) {
                try {
                    val fetched = runCatching {
                        coroutineScope {
                            needsFetch.chunked(5).map { chunk ->
                                chunk.map { item ->
                                    async(Dispatchers.IO) {
                                        try {
                                            val detailDoc = getDoc(item.url)
                                            val img = detailDoc.selectFirst("article .entry-content img")
                                                ?: detailDoc.selectFirst("article img")
                                                ?: detailDoc.selectFirst(".entry-content img")
                                            val imgUrl = img?.attr("src") ?: ""
                                            if (imgUrl.isNotBlank()) item to imgUrl else null
                                        } catch (_: Exception) { null }
                                    }
                                }.awaitAll().filterNotNull()
                            }.flatten()
                        }
                    }.getOrDefault(emptyList())

                    for ((item, imgUrl) in fetched) {
                        val idx = results.indexOfFirst { it.url == item.url }
                        if (idx >= 0) {
                            results[idx] = results[idx].copy(image = imgUrl)
                        }
                    }
                } catch (_: Exception) {}
            }
        } catch (e: Exception) {
            e.printStackTrace()
        }
        return results
    }
}
